fuzz.go

```diff
@@ -35,3 +38,98 @@ func Fuzz(data []byte) int {
 	}
 	return 0
 }
+
+// FuzzReadSubTreesWithProof can be used by go-fuzz to test creating a merkle
+// tree from cached subTrees and creating/proving a merkle proof on this tree.
+func FuzzReadSubTreesWithProof(data []byte) int {
+	// Use the first two bytes to determine the proof index.
+	if len(data) < 2 {
+		return -1
+	}
+	// Use the first two bytes to determine the proof index.
+	index := 256*uint64(data[0]) + uint64(data[1])
+	tree := New(sha256.New())
+	tree.SetIndex(index)
+	data = data[2:]
+
+	// 32 is the length of a sha256 hash and 4 bytes are used for the height of
+	// the subTree.
+	subTreeSize := 4 + 36
+	err := tree.readSubTrees(bytes.NewReader(data), subTreeSize)
+	if err != io.ErrUnexpectedEOF {
+		return 0
+	} else if err == io.ErrUnexpectedEOF && len(data) < subTreeSize {
+		return -1
+	}
+
+	// Create and verify the proof.
+	merkleRoot, proofSet, proofIndex, numLeaves := tree.Prove()
+	if len(proofSet) == 0 {
+		// proofIndex wasn't reached while creating proof.
+		return 0
+	}
+	if !VerifyProof(sha256.New(), merkleRoot, proofSet, proofIndex, numLeaves) {
+		panic("verification failed!")
+	}
+	// Output is more interesting when there is enough data to contain the
+	// index.
+	if uint64(len(data)) > uint64(subTreeSize)*index {
+		return 1
+	}
+	return 0
+}
+
+// FuzzReadSubTreesNoProof can be used by go-fuzz to test creating a merkle
+// tree from cached subTrees.
+func FuzzReadSubTreesNoProof(data []byte) int {
+	tree := New(sha256.New())
+
+	// 32 is the length of a sha256 hash and 4 bytes are used for the height of
+	// the subTree.
+	subTreeSize := 4 + 36
+	err := tree.readSubTrees(bytes.NewReader(data), subTreeSize)
+	if err != io.ErrUnexpectedEOF {
+		return 0
+	} else if err == io.ErrUnexpectedEOF && len(data) < subTreeSize {
+		return -1
+	}
+	if tree.head != nil && tree.Root() == nil {
+		panic("root shouldn't be nil for a non-empty tree")
+	}
+	// The data is better if it contains many subTrees.
+	if len(data) > 100*subTreeSize {
```

Comment:
100 is too high of a bar. I'd shoot for 2 trees. The fuzzer actually does a really good job on its own of deciding whether it needs to add more data or mix it up in some way, probably better to encourage it any time that it gets far enough to create a second tree.

If the fuzzer can get to 2 trees, leaving it running for a little extra time will let it get to dozens of trees.